File: sheet5/deviates.cpp
#include<stdlib.h>
#include<stdio.h>
#include<iostream>
#include<math.h>
#include<vector>
using namespace std;

// function headers

double distribution(double nu, double T);

double derivative(double nu, double T);

void build_dist(double limit, int num, double min, double max,
                vector<vector<double>> *acc, double T);
 
double bisection(double (*f)(double, double), double T, double upper, 
                 double lower, double eps);

bool criterium(double one, double two);


// main function

main(){

  // parameters

  double nu_min = 0;
  double nu_max = 5e+4;
  double alpha  = 0;
  double temperature    = 6e+3;
  double root_bisection = 0.0;
  double lower_bracket  = 9e+3;
  double upper_bracket  = 11e+3;
  double precision      = pow(10., -7.);
//  vector<double> point(2);
  vector<vector<double>> dist_values;
  int N = 10000; int i1 = 0; int i2 = 0;

  root_bisection = bisection(derivative, temperature, upper_bracket,
                             lower_bracket, precision);
//  cout << "root = " << root_bisection << endl;

  alpha = distribution(root_bisection, temperature); 
//  cout << "max(f(nu) = " << alpha << endl;

  build_dist(alpha, N, nu_min, nu_max, &dist_values, temperature);

//  for (i1 = 0; i1 < dist_values.size(); i1++) 
//    printf("%5.5f, %5.5f\n", dist_values[i1][0], dist_values[i1][1]);

  return 0;
}


// functions

double distribution(double nu, double T){
  return 8.*pow(M_PI, 2.)*pow(nu, 2.)*1./(exp(nu/T)-1.);
}

double derivative(double nu, double T){
  return distribution(nu, T) * (2./nu - 1./(T-T*exp(-nu/T)));
}

void build_dist(double limit, int num, double min, double max,
                vector<vector<double>> *acc, double T){
  srand(2);
  int i1 = 0; double albar = 0.0; // alpha_bar
  double nubar = 0.0; double fnubar = 0.0;
  vector<double> point(2);

  for (i1 = 0; i1 < num; i1++) {
    nubar = (double)rand()/(double)RAND_MAX * (max-min) + min;
   // cout << "nubar = " << nubar << endl;
    fnubar = distribution(nubar, T);
   // cout << "fnubar = " << fnubar << endl;
    albar = (double)rand()/(double)RAND_MAX * limit;
   // cout << "albar = " << albar << endl;
  
    if (albar > fnubar) continue;
    else {
      //point[0] = nubar; point[1] = albar;
      //acc->push_back(point);
      printf("%5.5f, %5.5f\n", nubar, albar);
    }
  }
}

/*
double rand_lim(double limit) {
  srand(2);
  double divisor = (double)RAND_MAX/(limit+1);
  double result = 0.0;
  do {result = (double)rand()/divisor} while (result > limit);
  return result;  
}
*/

double bisection(double (*f)(double, double), double T,
                 double upper, double lower, double eps){
  // result parameters
  double root  = 1.0; double root_prev = 1.0;
  double error = 5.0; int    count     = 0;

  // function evaluations
  double up  = 0.0; double low = 0.0; double mid = 0.0;

  // control
  // printf("upper: %5.3e\nlower: %5.3e\nprecn: %5.3e\n", upper, lower, eps);

  // bisection loop
  while (error > eps) {
    root_prev = root;
    low = f(lower, T);
    up  = f(upper, T);
    mid = f((lower+upper)/2., T);

    if (criterium(low, mid)) upper = (lower + upper) / 2.;
    else if (criterium(mid, up)) lower = (lower + upper)/2.;

    root = (lower + upper) / 2.;
    error = fabs((root/root_prev)-1);
    count += 1;
  } 

  printf("count, bisection: %d\n", count);
  return root;
}

bool criterium(double one, double two){
  if (one*two < 0) return true;
  else return false;
}

File: sheet5/deviates.py
import numpy  as np
import random as r
import math   as m
import matplotlib.pyplot as plt

def distribution(nu, T):
    return 8.*np.power(m.pi, 2.)*np.power(nu, 2.)*1./(np.exp(nu/T)-1.)

def derivative(nu, T):
    return distribution(nu, T)*(2./nu -1./(T-T*np.exp(nu/T)))

def crit(one, two):
    if one*two < 0:
        return True
    else:
        return False

def bisection(func, T, lower, upper, eps):
    error = 5.
    root = 5.
    while error > eps:
        root_prev = root
        low = func(lower, T)
        up = func(upper, T)
        mid = func((lower+upper)/2., T)

        if crit(low,mid):
            upper = (lower+upper)/2.
        elif crit(mid,up):
            lower = (lower+upper)/2.

        root = (lower+upper)/2.
        error = abs((root/root_prev)-1.)

    return root

def build_dist(limit, num, mini, maxi, T):
    dist = []

    for i1 in range (0, num):
        nubar  = r.uniform(mini, maxi)
        fnubar = distribution(nubar, T)
        albar  = r.uniform(0, limit)

        if albar > fnubar: 
            continue
        else:
            dist.append([nubar, albar])

    return dist

 
nu_min = 0
nu_max = 5e+4
temperature = 6e+3
lower  = 9e+3
upper  = 11e+3
precision   = 1e-7
number = 10000

root_dist = bisection(derivative, temperature, lower, upper, precision)
alpha = distribution(root_dist, temperature)
dist_list = build_dist(alpha, number, nu_min, nu_max, temperature)

nu, al = zip(*dist_list)

#for i1 in range(0, len(dist_list)):
#    nu.append(dist_list[i1][0])
#    al.append(dist_list[i1][1])

#new = np.hsplit(dist_list, 2)
# print(len(dist_list))
#print(dist_list)
plt.plot(nu, al, 'b+')
#plt.plot(new, 'r+')
plt.show()

